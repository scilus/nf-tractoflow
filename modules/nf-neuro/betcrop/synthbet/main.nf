process BETCROP_SYNTHBET {
    tag "$meta.id"
    label 'process_single'

    container "freesurfer/freesurfer:7.4.1"

    input:
    tuple val(meta), path(image), path(weights) /* optional, input = [] */

    output:
    tuple val(meta), path("*__bet_image.nii.gz"), emit: bet_image
    tuple val(meta), path("*__brain_mask.nii.gz"), emit: brain_mask
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def gpu = task.ext.gpu ? "--gpu" : ""
    def border = task.ext.border ? "-b " + task.ext.border : ""
    def nocsf = task.ext.nocsf ? "--no-csf" : ""
    def model = "$weights" ? "--model $weights" : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    mri_synthstrip -i $image --out ${prefix}__bet_image.nii.gz --mask ${prefix}__brain_mask.nii.gz $gpu $border $nocsf $model

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        Freesurfer: \$(mri_convert -version | grep "freesurfer" | sed -E 's/.* ([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    touch ${prefix}__bet_image.nii.gz
    touch ${prefix}__brain_mask.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        Freesurfer: \$(mri_convert -version | grep "freesurfer" | sed -E 's/.* ([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/')
    END_VERSIONS

    function handle_code () {
    local code=\$?
    ignore=( 1 )
    exit \$([[ " \${ignore[@]} " =~ " \$code " ]] && echo 0 || echo \$code)
    }
    trap 'handle_code' ERR

    mri_synthstrip -h
    """
}
