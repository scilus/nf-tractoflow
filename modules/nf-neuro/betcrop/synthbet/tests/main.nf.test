nextflow_process {

    name "Test Process BETCROP_SYNTHBET"
    script "../main.nf"
    process "BETCROP_SYNTHBET"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "betcrop"
    tag "betcrop/synthbet"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    test("betcrop - synthbet") {
        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "freesurfer.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/anat_image.nii.gz"),
                        []
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("betcrop - synthbet - stub-run") {
        options "-stub-run"
        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "freesurfer.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/anat_image.nii.gz"),
                        []
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
