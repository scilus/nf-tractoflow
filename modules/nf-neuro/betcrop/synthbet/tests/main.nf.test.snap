{
    "betcrop - synthbet - stub-run": {
        "content": [
            [
                "versions.yml:md5,c639461870ca534b5105f61d672f740f"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.1"
        },
        "timestamp": "2024-12-12T10:54:37.644838"
    },
    "betcrop - synthbet": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bet_image.nii.gz:md5,6f5eb365a11dfc52b8c80cd7bcee7b9d"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__brain_mask.nii.gz:md5,52871114b59d09dca864cb0ea4f12832"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c639461870ca534b5105f61d672f740f"
                ],
                "bet_image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bet_image.nii.gz:md5,6f5eb365a11dfc52b8c80cd7bcee7b9d"
                    ]
                ],
                "brain_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__brain_mask.nii.gz:md5,52871114b59d09dca864cb0ea4f12832"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c639461870ca534b5105f61d672f740f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-25T18:19:01.459548926"
    }
}